Dashboard: Fix Editing a panel with auto-refresh enabled auto-refresh in the edit screen

axelavargas opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

When a Dashboard has the auto-refresh option set, then the panel edit screen also had it, it was causing unexpected refresh for the users if, for example, they were setting a query. 

**Approach**

The approach in this PR:
- **Hide auto-refresh option(The interval one) on panel edit**, the reason behind hiding the option instead of not having the auto-refresh enabled by default is to avoid keeping track of the auto-refresh values changed per panel. (currently, we only support global auto-refresh that apply to all the panels)

Before:

![Screenshot from 2021-05-17 15-03-51](https://user-images.githubusercontent.com/239999/118493373-543cb300-b721-11eb-8c0a-6f97b892738e.png)


**This is how will look like with the option hidden, the refresh (sync button) is there:**

![Screenshot from 2021-05-17 15-00-51](https://user-images.githubusercontent.com/239999/118492941-dd9fb580-b720-11eb-96fb-c82f4cdf30f7.png)


**Which issue(s) this PR fixes**:


<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #21764

**Special notes for your reviewer**:

I am lacking some context in the project, so it might be worth thinking on missing edge cases where this solution does not work

axelavargas committed:
Dashboard: Fix editing panel with auto-refresh

- Add default hidden and disabled auto-refresh on editing panel
- Extend unit test for DashboardModel
- Add unit test DashNavTimeControls component

axelavargas committed:
Test: Extend DashboardModel.initEditPanel unit test

hugohaggmark commented on public/app/features/dashboard/components/DashNav/DashNavTimeControls.tsx:
```diff
@@ -88,6 +88,7 @@ export class DashNavTimeControls extends Component<Props> {
     const timePickerValue = getTimeSrv().timeRange();
     const timeZone = dashboard.getTimezone();
     const styles = getStyles();
+    const hideRefreshPicker = !dashboard.panelInEdit?.isEditing;
```
Hmm, can we reverse the naming of this variable? Shouldn't it be showRefreshPicker?

axelavargas replied:
Thanks for the suggestion :D, I also did a small fix here as well,  this component also had the `refresh/sync` button that needs to be always present https://github.com/grafana/grafana/pull/34128/commits/02f4f35075001434564b9a2244a10706936f7b8c#diff-aac52f3778332268268f492b861c176a728fa4e7f05a4a5574228ec00920482aR91

hugohaggmark replied:
Looks great 👍

hugohaggmark commented on public/app/features/dashboard/state/DashboardModel.test.ts:
```diff
@@ -849,8 +859,15 @@ describe('exitPanelEditor', () => {
+
     describe('and there is a change that affects all panels', () => {
-      it('then startRefresh is not called', () => {
+      it('then startRefresh is called', () => {
```
Nice catch!

hugohaggmark commented on public/app/features/dashboard/state/DashboardModel.test.ts:
```diff
@@ -7,8 +7,16 @@ import { createAdHocVariableAdapter } from '../../variables/adhoc/adapter';
+import { setTimeSrv, TimeSrv } from '../services/TimeSrv';
 
 jest.mock('app/core/services/context_srv', () => ({}));
+
+const timeSrvMock = ({
+  setAutoRefresh: jest.fn(),
+} as unknown) as TimeSrv;
+
+setTimeSrv(timeSrvMock);
+
```
nit: maybe the `setTimeSrv(timeSrvMock);` should be set within the `getTestContext` functions so it's not set in the global scope?

axelavargas replied:
indeed, a good idea 🙌🏾 https://github.com/grafana/grafana/pull/34128/commits/02f4f35075001434564b9a2244a10706936f7b8c#diff-0338d437d3a32fb5066ab71687aecdaf0229a778cde30e595af9ff1a39f152d1R823 🚀

hugohaggmark commented on public/app/features/dashboard/state/DashboardModel.ts:
```diff
@@ -356,6 +359,8 @@ export class DashboardModel {
   }
 
   initEditPanel(sourcePanel: PanelModel): PanelModel {
+    this.previousAutoRefresh = this.refresh;
+    getTimeSrv().setAutoRefresh('');
```
nit: maybe this logic belongs to `TimeSrv`,  we could have an `pauseAutoRefresh` and `resumeAutoRefresh` function that handles this logic. But at the same time it feels strange that TimeSrv should mutate dashboard state directly.

hugohaggmark replied:
> nit: maybe this logic belongs to `TimeSrv`, we could have an `pauseAutoRefresh` and `resumeAutoRefresh` function that handles this logic. But at the same time it feels strange that TimeSrv should mutate dashboard state directly. (But it already does)

axelavargas replied:
I agree with you,  this logic belongs to the `TimeSrv`, I  did a small refactor in the code https://github.com/grafana/grafana/pull/34128/commits/e553524da6f8cbaea5c20c12e8f270ccbade62b8 and now we are controlling the value of the `previousAutoRefresh` in the `TimeSrv`

hugohaggmark replied:
Beautiful 🎉

hugohaggmark commented on public/app/features/dashboard/components/DashNav/DashNavTimeControls.test.tsx:
```diff
@@ -0,0 +1,49 @@
+  beforeEach(() => {
+    const json = {
+      panels: [
+        {
+          datasource: null,
+          gridPos: {
+            h: 3,
+            w: 24,
+            x: 0,
+            y: 8,
+          },
+          id: 1,
+          type: 'welcome',
+        },
+      ],
+      refresh: '',
+      templating: {
+        list: [],
+      },
+    };
+    dashboardModel = getDashboardModel(json);
+  });
+
+  it('renders RefreshPicker in panel view', () => {
+    const container = render(
+      <DashNavTimeControls dashboard={dashboardModel} onChangeTimeZone={jest.fn()} key="time-controls" />
+    );
+    expect(container.queryByLabelText(/RefreshPicker run button/i)).toBeInTheDocument();
+  });
+
+  it('should not render RefreshPicker in panel edit', () => {
+    const panel: any = { destroy: jest.fn(), isEditing: true };
+    dashboardModel.startRefresh = jest.fn();
+    dashboardModel.panelInEdit = panel;
+    const container = render(
+      <DashNavTimeControls dashboard={dashboardModel} onChangeTimeZone={jest.fn()} key="time-controls" />
+    );
+    expect(container.queryByLabelText(/RefreshPicker run button/i)).not.toBeInTheDocument();
+  });
+});
```
Skilled work adding these tests 🎉

hugohaggmark reviewed (approved):
Skilled work on this and works as expected. Outstanding that you added tests, I left some small comments for consideration.

torkelo reviewed (commented):
Great job! beyond the small comments  Hugo added, this looks great!

axelavargas committed:
Merge branch 'main' into axelav/prevent-auto-refresh-edit-mode

axelavargas committed:
Dashboard: Fix refresh button should be available on Panel Edit

- hide on panel edit interval auto-refresh option
- show on panel edit refresh button
- update test

axelavargas committed:
Dashboard: Refactor auto-refresh functionality (pause and resumw), move logic to TimeSrv service

- Move pause and resume auto-refresh functionality to TimeSrv
- Add and modify exisiting test

hugohaggmark commented:
@axelavargas feel free to `Squash and merge` when you're finished with this PR 👍

hugohaggmark commented:
![](https://media2.giphy.com/media/l7fdqmHQ1jCg2HzQlx/giphy.gif)
